Modules/hmacmodule.c

```diff
@@ -0,0 +1,277 @@
+                            Py_buffer *data)
+/*[clinic end generated code: output=cf977eed9c59ed3b input=ce59d1ddd77c0624]*/
+{
+    HACL_HMAC_COMPUTE_HASH(Hacl_HMAC_compute_sha3_256, 32, key, data);
+}
+
+/*[clinic input]
+@critical_section
+_hmac.compute_sha3_384
+
+    key: Py_buffer
+    data: Py_buffer
+    /
+
+[clinic start generated code]*/
+
+static PyObject *
+_hmac_compute_sha3_384_impl(PyObject *module, Py_buffer *key,
+                            Py_buffer *data)
+/*[clinic end generated code: output=3f576e31d4d05f35 input=f4bca88551693caa]*/
+{
+    HACL_HMAC_COMPUTE_HASH(Hacl_HMAC_compute_sha3_384, 48, key, data);
+}
+
+/*[clinic input]
+@critical_section
+_hmac.compute_sha3_512
+
+    key: Py_buffer
+    data: Py_buffer
+    /
+
+[clinic start generated code]*/
+
+static PyObject *
+_hmac_compute_sha3_512_impl(PyObject *module, Py_buffer *key,
+                            Py_buffer *data)
+/*[clinic end generated code: output=238126dcba98fda2 input=2f98f302c64eca64]*/
+{
+    HACL_HMAC_COMPUTE_HASH(Hacl_HMAC_compute_sha3_512, 64, key, data);
+}
+
+/*[clinic input]
+@critical_section
+_hmac.compute_blake2s_32
+
+    key: Py_buffer
+    data: Py_buffer
+    /
+
+[clinic start generated code]*/
+
+static PyObject *
+_hmac_compute_blake2s_32_impl(PyObject *module, Py_buffer *key,
+                              Py_buffer *data)
+/*[clinic end generated code: output=72a8231623e4ccf9 input=0be9099b69bcd9e7]*/
+{
+    HACL_HMAC_COMPUTE_HASH(Hacl_HMAC_compute_blake2s_32, 32, key, data);
+}
+
+/*[clinic input]
+@critical_section
+_hmac.compute_blake2b_32
+
+    key: Py_buffer
+    data: Py_buffer
+    /
+
+[clinic start generated code]*/
+
+static PyObject *
+_hmac_compute_blake2b_32_impl(PyObject *module, Py_buffer *key,
+                              Py_buffer *data)
+/*[clinic end generated code: output=ea083dfa29679029 input=aecba54a3e2dff72]*/
+{
+    HACL_HMAC_COMPUTE_HASH(Hacl_HMAC_compute_blake2b_32, 32, key, data);
+}
+
+static PyMethodDef hmacmodule_methods[] = {
+    _HMAC_COMPUTE_MD5_METHODDEF
+    _HMAC_COMPUTE_SHA1_METHODDEF
+    _HMAC_COMPUTE_SHA2_224_METHODDEF
+    _HMAC_COMPUTE_SHA2_256_METHODDEF
+    _HMAC_COMPUTE_SHA2_384_METHODDEF
+    _HMAC_COMPUTE_SHA2_512_METHODDEF
+    _HMAC_COMPUTE_SHA3_224_METHODDEF
+    _HMAC_COMPUTE_SHA3_256_METHODDEF
+    _HMAC_COMPUTE_SHA3_384_METHODDEF
+    _HMAC_COMPUTE_SHA3_512_METHODDEF
+    _HMAC_COMPUTE_BLAKE2S_32_METHODDEF
+    _HMAC_COMPUTE_BLAKE2B_32_METHODDEF
+    {NULL, NULL, 0, NULL}
+};
+
+static struct PyModuleDef_Slot hmacmodule_slots[] = {
+    {Py_mod_multiple_interpreters, Py_MOD_PER_INTERPRETER_GIL_SUPPORTED},
+    {Py_mod_gil, Py_MOD_GIL_NOT_USED},
+    {0, NULL}
+};
+
+static struct PyModuleDef _hmacmodule = {
+    PyModuleDef_HEAD_INIT,
+    .m_name = "_hmac",
+    .m_size = 0,
+    .m_methods = hmacmodule_methods,
+    .m_slots = hmacmodule_slots,
+};
+
+PyMODINIT_FUNC
+PyInit__hmac(void)
+{
+    return PyModuleDef_Init(&_hmacmodule);
```

Comment:
(This is primarily a note for myself when this is closer to being ready.)

Do modules without a `m_traverse` get to use DRC to ease contention (which will be especially bad here because of the critical sections)? If not, force the module to do that, or consider making this immortal somehow. (I'll follow up on this--it will take some work, because I don't think we're doing it anywhere else in the stdlib.)

Reply:
(For others: DRC means deferred reference count). This is something I also wondered, namely should I explicitly call `Py_VISIT(Py_TYPE(mod))` even if I don't have a state or not? I assume *not* because that's what multiple modules with `m_size = 0` do as well. I don't know where to find an exact reference on that behaviour though.